lib/internal/streams/duplexpair.js

```diff
@@ -39,13 +38,11 @@ class DuplexSocket extends Duplex {
   }
 }
 
-class DuplexPair {
-  constructor() {
-    this.socket1 = new DuplexSocket();
-    this.socket2 = new DuplexSocket();
-    this.socket1[kOtherSide] = this.socket2;
-    this.socket2[kOtherSide] = this.socket1;
-  }
+function duplexPair(options) {
+  const side0 = new DuplexSide(options);
+  const side1 = new DuplexSide(options);
+  side0[kOtherSide] = side1;
+  side1[kOtherSide] = side0;
```

Comment:
I think it's better to restrict direct access for those properties by encapsulation, so they can't be modified directly from outside the class.

Reply:
I agree with this... I may have been under the impression that you can't acquire the value of kOtherSide from outside the file (but actually you can enumerate these). Would you suggest a WeakMap instead?

Reply:
> Would you suggest a WeakMap instead?

well, it would be a bit better with WeakMap, something like that:
```js
const privateData = new WeakMap();

class DuplexSide extends Duplex {
  constructor() {
    privateData.set(this, {
      [kCallback]: null,
      [kOtherSide]: null,
    });
  }

  setOtherSide(otherSide) {
    const data = privateData.get(this);
    if (data[kOtherSide]) {
      throw new Error('readonly')
    }
    data[kOtherSide] = otherSide;
  }
}
```
however, even that approach cannot guarantee truly private properties that can't be directly accessed or modified, I would say it's only hope on convention. WeakMap provide a level of privacy and make it difficult for external code to accidentally interfere, but they do not make the properties completely immutable.

I suggest using the redefinition of `set` method together with `Proxy` object for that, to make sure of immutable:
```js
const handler = {
  set(target, prop, value) {
    if (prop in target && target[prop] !== null) {
      throw new Error('Property is already set.')
    } else {
      target[prop] = value
      return true
    }
  }
}

class DuplexSocket {
  constructor() {
    this[kCallback] = null
    this[kOtherSide] = null
    return new Proxy(this, handler)
  }
}
```
🗒️ `target[prop] !== null` condition needs to handle first assign properties as null in the constructor.

📓 specs:
```js
  test('allows assign kCallback.', async () => {
    const socket = new DuplexSocket()
    socket[kCallback] = 'value kCallback 1'
    expect(socket[kCallback]).toStrictEqual('value kCallback 1')
  })

  test('forbids reassign kCallback.', async () => {
    const socket = new DuplexSocket()
    socket[kCallback] = 'value kCallback 1'
    expect(() => {
      socket[kCallback] = 'value kCallback 2'
    }).toThrow(new Error('Property is already set.'))
    expect(socket[kCallback]).toStrictEqual('value kCallback 1')
  })

  test('allows assign kOtherSide.', async () => {
    const socket = new DuplexSocket()
    socket[kOtherSide] = 'value kOtherSide 1'
    expect(socket[kOtherSide]).toStrictEqual('value kOtherSide 1')
  })

  test('forbids reassign kOtherSide.', async () => {
    const socket = new DuplexSocket()
    socket[kCallback] = 'value kOtherSide 1'
    expect(() => {
      socket[kCallback] = 'value kOtherSide 2'
    }).toThrow(new Error('Property is already set.'))
    expect(socket[kCallback]).toStrictEqual('value kOtherSide 1')
  })
```

Reply:
Let's not use WeakMaps for this, they are incredibly expensive from a GC standpoint.